### PR TITLE
support static link

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang @disksing @zhangjinpeng1987 PTAL

### CODE REVIEW

**Comment by ngaut** [comment]
Good job.

**Comment by ngaut** [comment]
LGTM

**Comment by c4pt0r** [comment]
Well done! LGTM

File: librocksdb_sys/build.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    [[ "$hash" == "$3" ]] || panic $2: hash not correct, expect $3, got $hash
+}
+
+function compile_z() {
+    if [[ -f libz.a ]]; then
+        return
+    fi
+
+    rm -rf zlib-1.2.8
+	download http://zlib.net/zlib-1.2.8.tar.gz zlib-1.2.8.tar.gz 44d667c142d7cda120332623eab69f40
```
Comment (siddontang) [inline librocksdb_sys/build.sh]: alignment?

**Comment by siddontang** [comment]
Can we supply a ready compiled rocksdb lib for different platforms so that here can use it directly, no need to install g++ and compile all never?

**Comment by BusyJay** [comment]
I don't think it's our job to do this. And the compilation won't be executed if users compile it without ROCKSDB_SYS_STATIC flag, which will fallback to dynamic link.

File: librocksdb_sys/build.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    if [[ -f librocksdb.a ]]; then
+        return
+    fi
+
+    echo building rocksdb
+    rm -rf rocksdb-4.8
+    download https://github.com/facebook/rocksdb/archive/v4.8.tar.gz rocksdb-4.8.tar.gz c7e1204cc6072d1a3e47817834b31c63
+    tar xf rocksdb-4.8.tar.gz
+    cd rocksdb-4.8
+    EXTRA_CFLAGS="-fPIC -I./zlib-1.2.8 -I./bzip2-1.0.6 -I./snappy-1.1.1 -I./lz4-r127/lib" EXTRA_CXXFLAGS="$EXTRA_CFLAGS" PORTABLE=1 make static_lib
```
Comment (siddontang) [inline librocksdb_sys/build.sh]: need PORTABLE?

Comment (BusyJay) [inline librocksdb_sys/build.sh]: Yes.
>By default the binary we produce is optimized for the platform you're compiling on (-march=native or the equivalent). If you want to build a portable binary, add 'PORTABLE=1' before your make commands, like this: PORTABLE=1 make static_lib

**Comment by siddontang** [comment]
Can https://github.com/rust-lang/rust/issues/15420 solve static c++ linking?

File: librocksdb_sys/build.sh (hunk 2 above)
Comment (siddontang) [inline librocksdb_sys/build.sh]: We don't need PORTABLE, e,g, if we build tikv + rocksdb in centos, we only use it in centos, won't use it in ubuntu or mac os x. 
So it is better to build the native rocksdb.

Comment (BusyJay) [inline librocksdb_sys/build.sh]: I'm not sure whether the platform here includes hardware spec, like instruction of cpu and so on.

Comment (siddontang) [inline librocksdb_sys/build.sh]: Native has better performance.
If two machines has different hardware, I think we may not run tikv too even using PORTABLE rocksdb. 

/cc @ngaut

**Comment by BusyJay** [comment]
I can't find related code in the latest master anymore.

**Comment by siddontang** [comment]
```
✗ rustc -C help

Available codegen options:

    -C                ar=val -- tool to assemble archives with
    -C            linker=val -- system linker to link outputs with
    -C         link-args=val -- extra arguments to pass to the linker (space separated)
```

Maybe we can use link-args ?

**Comment by BusyJay** [comment]
`-static-libstdc++` works only when using c++ compiler. But rustc seems using c compiler.

File: librocksdb_sys/build.sh [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    download https://github.com/Cyan4973/lz4/archive/r131.tar.gz lz4-r131.tar.gz 42b09fab42331da9d3fb33bd5c560de9
+    tar xvzf lz4-r131.tar.gz
+    cd lz4-r131/lib
+    make CFLAGS='-fPIC' all
+    mv liblz4.a ../../
+    cd ../..
+}
+
+function compile_rocksdb() {
+    if [[ -f librocksdb.a ]]; then
```
Comment (siddontang) [inline librocksdb_sys/build.sh]: how to upgrade rocksdb?

Comment (BusyJay) [inline librocksdb_sys/build.sh]: `cargo clean`.

Comment (siddontang) [inline librocksdb_sys/build.sh]: so if we use `cargo update rocksdb`, we should use `cargo clean` then rebuild?

Comment (BusyJay) [inline librocksdb_sys/build.sh]: Yes.

Comment (siddontang) [inline librocksdb_sys/build.sh]: use rocksdb master now

File: librocksdb_sys/build.sh (hunk 2 above)
Comment (siddontang) [inline librocksdb_sys/build.sh]: I prefer we can let user determine whether use PORTABLE=1 or not outer, like other flags. 
If user has same platform, hardware, no need PORTABLE.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**